### ISSUE TITLE
Old version of build.gradle used

### ISSUE THREAD

**Issue by Vampire** [gradle/gradle#7992]
Unfortunately I have no idea how to reproduce this, but I just had a very strange behavior that I feel the urge to report anyway.

I modified a build script and wanted to build, but Gradle uses the old version of the build script.
I can see this from the build result and also from debugging into relevant code and seeing that the script is the old version.

Neither adding `--recompile-scripts`, nor deleting `<project>/.gradle` or restarting the daemon had any effect, it was always the old version of the script used.

The only way to make it use the new version of the script was to modify it. I added an arbitrary space character, then the new version was used. Now if I remove that space character again, again the old version of the build script is used, so I guess it is something in `~/.gradle` that is borked and I find no way to fix it besides maybe deleting things in `~/.gradle` which I didn't do yet to maybe be able to debug the problem further if you can tell me where to look at.

**Comment by Vampire** [comment]
Yep, it is `$GRADLE_HOME\caches\4.10.2\scripts-remapped\build_53p7oouiun0h8o0luzyzt4k4b\8dyv8hf0rhmx8fdyjixjjyf42\proj4968aefb9f2665c3148724a8fb2f7e6a\classes\build_53p7oouiun0h8o0luzyzt4k4b.class` which has the old content, while the `TextResourceScriptSource` and even the `CachingScriptSource` have the correct code.

**Comment by Vampire** [comment]
It seems that somehow the old compiled script was stored in the cache with the hash of the new version of the script, I'd say that is the base bug, while I have no idea how to reproduce it, but I already got this behavior before, just didn't report it and didn't care too much as I had to change the script further anyway.

But I still wonder that even `--recompile-scripts` does not fix it.

**Comment by Vampire** [comment]
Oh, `--recompile-scripts` actually **does** help, but only if you first restart the daemon.
I guess there is some in-memory cache that has the compiled script and if that is not filled, it looks for the file-based cache if the option is not given.

But why is `--recompile-scripts` deprecated?
Is there a replacement?
In my case I would have had to find out which cache is used and delete it manually or delete the whole `$GRADLE_HOME\caches\4.10.2\scripts-remapped` directory **and** restart the daemon so that the in-memory cache is not used, but you have to know all those internals.
I would open a separate bug because of the wrong `--recompile-scripts` behavior if it were not deprecated. :-(

So besides this question, the main issue seems to be that somehow the compiled script got stored in the wrong cache directory. My guess would be, that the compilation started, then I changed the script while compilation was running and then the cache dir was calculated based on the state of the file when compilation was finished or something like that. This would at least be a viable explanation.

**Comment by oehme** [comment]
We have all kinds of guards against this situation, including:

-  a check that verifies the hash that was used to compile the script against the current hash of the source
- getting the content of the script once and using the cached value both for computing its cache key and for compilation, so changes during the build are ignored

We only know of one rare condition when stale scripts are used:  A script file was changed, but kept both length and lastModifiedTime intact. We've only seen this with some automated scripts that deliberately reset the modification timestamp though. We don't consider this scenario common enough to warrant a command line option.

If you can reproduce a different cause for stale scripts, please let me know.

**Comment by Vampire** [comment]
The changes were in neither cases automated, but manual editing within IntelliJ. The size should not have been the same and the last modified date most probably also not.
As I said it happened at least three times to me, but I cannot provide a reliable reproduction recipe.

**Comment by oehme** [comment]
I took another look and there actually is [special-casing](https://github.com/gradle/gradle/blob/d4ef7d22e0b9cb8576ad5cbe41c425d9d1919d6e/subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/DefaultScriptSourceHasher.java#L25) for file-based script sources which hashes the underlying file instead of using the text we have in memory. This is most likely responsible for any discrepancies between the hash and the compiled file. I don't know why we did this, the git history doesn't tell. I'll run a fix through our pipeline to see if something breaks.
